routesrv/routesrv.go

```diff
@@ -0,0 +1,154 @@
+// This call does not start data source updates automatically. Kept routes
+// will stay in an uninitialized state, till StartUpdates is called and
+// in effect data source is queried and routes initialized/updated.
+func New(opts Options) (*RouteServer, error) {
+	rs := &RouteServer{}
+
+	opentracingOpts := opts.OpenTracing
+	if len(opentracingOpts) == 0 {
+		opentracingOpts = []string{"noop"}
+	}
+	tracer, err := tracing.InitTracer(opentracingOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	b := &eskipBytes{tracer: tracer}
+	bs := &eskipBytesStatus{b: b}
+	handler := http.NewServeMux()
+	handler.Handle("/health", bs)
+	handler.Handle("/routes", b)
+	handler.Handle("/metrics", promhttp.Handler())
+	rs.server = &http.Server{Addr: opts.Address, Handler: handler}
+
+	dataclient, err := kubernetes.New(kubernetes.Options{
+		KubernetesInCluster:               opts.KubernetesInCluster,
+		KubernetesURL:                     opts.KubernetesURL,
+		ProvideHealthcheck:                opts.KubernetesHealthcheck,
+		ProvideHTTPSRedirect:              opts.KubernetesHTTPSRedirect,
+		HTTPSRedirectCode:                 opts.KubernetesHTTPSRedirectCode,
+		IngressClass:                      opts.KubernetesIngressClass,
+		RouteGroupClass:                   opts.KubernetesRouteGroupClass,
+		ReverseSourcePredicate:            opts.ReverseSourcePredicate,
+		WhitelistedHealthCheckCIDR:        opts.WhitelistedHealthCheckCIDR,
+		PathMode:                          opts.KubernetesPathMode,
+		KubernetesNamespace:               opts.KubernetesNamespace,
+		KubernetesEastWestRangeDomains:    opts.KubernetesEastWestRangeDomains,
+		KubernetesEastWestRangePredicates: opts.KubernetesEastWestRangePredicates,
+		DefaultFiltersDir:                 opts.DefaultFiltersDir,
+		BackendNameTracingTag:             opts.OpenTracingBackendNameTag,
+		OnlyAllowedExternalNames:          opts.KubernetesOnlyAllowedExternalNames,
+		AllowedExternalNames:              opts.KubernetesAllowedExternalNames,
+	})
+	if err != nil {
+		return nil, err
+	}
+	rs.poller = &poller{
+		client:  dataclient,
+		timeout: opts.SourcePollTimeout,
+		b:       b,
+		quit:    make(chan struct{}),
+		tracer:  tracer,
+	}
+
+	rs.wg = &sync.WaitGroup{}
+
+	return rs, nil
+}
+
+// StartUpdates starts the data source polling process.
+func (rs *RouteServer) StartUpdates() {
+	rs.wg.Add(1)
+	go rs.poller.poll(rs.wg)
+}
+
+// StopUpdates stop the data source polling process.
+func (rs *RouteServer) StopUpdates() {
+	rs.poller.quit <- struct{}{}
+}
+
+// ServeHTTP serves kept eskip-formatted routes under /routes
+// endpoint. Additionally it provides a simple health check under
+// /health and Prometheus-compatible metrics under /metrics.
+func (rs *RouteServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	rs.server.Handler.ServeHTTP(w, r)
+}
+
+func newShutdownFunc(rs *RouteServer) func(delay time.Duration) {
+	once := &sync.Once{}
+	rs.wg.Add(1)
```

Comment:
Hm, not sure why we need to tick wg here...

Reply:
Ah, I see, we reuse it wait for shutdown completion

Reply:
Indeed. It is a simple convenience that helps treating both shutdown scenarios the same way and spares additional syncing structures, however, I admit, is harder to read / express the purpose.